Add BMP support

Hakerh400 opened the pull request:
Fix #953

- [x] Update changelog
- [x] Add test

zbjornson commented on src/Image.cc:
```diff
@@ -1223,6 +1230,78 @@ Image::loadSVG(FILE *stream) {
+
+  if (parser.getStatus() != BMPParser::Status::OK) {
+    errorInfo.reset();
+    errorInfo.cerrno = 0;
```
Minor: don't need to set this to 0; `reset()` does that.

Hakerh400 replied:
fixed

Hakerh400 commented:
The feature request asking for BMP support has been open for more than a year, while implementing a simple BMP parser is trivial by following the specification. The pixel formats and compression algorithms it supports are the most common used ones, so I think it covers about 95% BMP files that average user may try to load. If anyone opens an issue with unsupported BMP, we can easily implement a new pixel format/compression method.

Hakerh400 commented on src/Image.cc:
```diff
@@ -1223,6 +1230,78 @@ Image::loadSVG(FILE *stream) {
+    return CAIRO_STATUS_READ_ERROR;
+  }
+
+  uint8_t *buf = (uint8_t*)malloc(s.st_size);
```
fixed

LinusU commented:
Having our own BMP implementation rubs me the wrong way a bit, but if there is no established external library than I guess it's the way to go 🤔

zbjornson commented:
I'm not opposed to having our own if it addresses our needs; I was just curious if it needs to be reviewed closely or if it's already in use elsewhere and how compatible it is. This parser is from-scratch, I take it?

https://github.com/ArashPartow/bitmap is the header-only library I came across that supports several color conversions. It does more than we need (bitmap library *plus* a graphics library), and would need modification to accept a memory source (instead of only a file source -- e.g. make it take an istream or array).

Hakerh400 commented:
> if it needs to be reviewed closely or if it's already in use elsewhere

It's not in use elsewhere. You can review it closely if you want.

> I came across a header-only library that supports more pixel formats

It supports only 24-bit images. The image from issue #953 is 1-bit (monochrome).

LinusU commented:
I have used a JS implementation of a BMP parser inside [`decode-ico`](https://www.npmjs.com/package/decode-ico) for long over a year in production, I just extracted the BMP parser into it's own package: [`decode-bmp`](https://www.npmjs.com/package/decode-bmp)

Potentially we could use that implementation, I'm not sure how much faster or slower it is than a C++ implementation though 🤔 Would be very interesting to see some benchmarks ☺️

Hakerh400 commented:
> Would be very interesting to see some benchmarks

```js
const fs = require('fs');
const {Image} = require('canvas');
const decodeBmp = require('decode-bmp');

const source = fs.readFileSync('10000x10000-monochrome.bmp');
const img = new Image;

console.time('BMPParser');
img.src = source; // Synchronous
console.timeEnd('BMPParser');

console.time('decode-bmp');
decodeBmp(source);
console.timeEnd('decode-bmp');
```

```console
> node test.js
BMPParser: 812.537ms
decode-bmp: 5899.919ms
```

Hakerh400 commented:
Update: added support for another common BMP header.
While we are awaiting @zbjornson's detailed review, here is the parser comparison:

||ArashPartow/bitmap|decode-bmp|BMPParser|
| :--- | :---: | :---: | :---: |
|Supports 1-bit images|:x:|:heavy_check_mark:|:heavy_check_mark:|
|Supports 32-bit images|:x:|:heavy_check_mark:|:heavy_check_mark:|
|Supports pre-multiplied alpha|:x:|:x:|:heavy_check_mark:|
|Fast|:heavy_check_mark:|:x:|:heavy_check_mark:|
|Easy to add new features|:x:|:heavy_check_mark:|:heavy_check_mark:|
|Parses [the minimal BMP](https://raw.githubusercontent.com/Hakerh400/node-canvas/77bef4d2540dfaca73623f9ee3aaab45a5a94eaf/test/fixtures/bmp/min.bmp)|:heavy_check_mark:|:x:|:heavy_check_mark:|

---

Feel free to correct me if anything in the table is incorrect.

Hakerh400 commented:
Thanks for the review.
Added fixes and also support for negative height.
After reading the specification for the second time, it seems that 0 is not valid value for width or height, and also width may not be negative.

> I don't think you want to be using std::byte anywhere. (a) It requires C++17 and we can't use past C++14 right now for compatibility. (b) Doing anything but bitwise ops on a byte is undefined behavior (or an implicit cast?). uint8_t is probably what you want.

It's actually `uint8_t`
In the `BMPParser.h` header file there is type definition: `typedef uint8_t byte;`

> You've called setErr (good) but you're not breaking the loops where this is used. You're at least returning a 0 instead of uninitialized memory or overrunning the buffer though.

Yes, that is correct. I added check before entering the loop to ensure that image data 1) doesn't overlap with the header and 2) doesn't exceed the buffer size. Also removed checks from the loop.

> since you're supporting no compression and bitfield compression, you should know exactly how many bytes you need the buffer to contain before you start the decoding loop.

The specification says that if compression is used, the valid image data size must be present in the header. So, all checks can be moved out of the loop even if compression is used.

zbjornson commented on src/bmp/BMPParser.h:
```diff
@@ -10,10 +10,9 @@
-
 namespace BMPParser{
   typedef uint8_t byte;
+  typedef std::string string;
```
Ah, I didn't see the `uint8_t byte` typedef in my initial review; I thought you were using `std::byte`.

This string typedef now creates an ambiguous symbol (`BMPParser::string` vs. `std::string` -- compilation error with MSVC but not GCC apparently).

Do you need either typedef? It'd be clearer without them I think, and avoid the above bug.

Hakerh400 replied:
Removed both typedefs.
But that is weird, I compiled with the latest msvc and didn't get even a warning.

zbjornson commented on src/bmp/BMPParser.cc:
```diff
@@ -214,14 +240,18 @@ void Parser::parse(byte *buf, int bufSize, byte *format){
-  for(int y = h - 1; y != -1; y--){
+  // Don't perform overrun checks in the loop
+  #undef GET_METHOD
+  #define GET_METHOD getc
```
In the category of macros-are-evil... I think this would be better done as a template parameter, something like:

```cpp
// BMPParser.h
template <typename T, bool skipcheck = false> T get();

// BMPParser.cc
template <typename T, bool skipcheck> T get() {
  if (!skipcheck) CHECK_OVERRUN(sizeof(T), T);
  // ...
}
```

Then however you prefer ... new `#define`s for like `U1UC` ("U1 unchecked"), or `get<uint8_t, true>()`,

zbjornson commented on src/bmp/BMPParser.h:
```diff
@@ -10,10 +10,9 @@
 #include <string>
 #include <cassert>
```
Is `iostream` used?

Looks like the `cassert` include can be moved to the .cc file.

(`string` should be in both, since it's used in both.)

zbjornson commented on src/bmp/BMPParser.cc:
```diff
@@ -0,0 +1,403 @@
+  imgd = nullptr;
+
+  op = "";
+  err = "";
```
Up to you, but this entire ctor body could be done with default member initializers (i.e. in the header), and then you can do `Parser() = default`. Would save some lines.

```cpp
class Parser{
public:
  Parser() = default; 
  // ...
private:
  Status status = Status::EMPTY;
  // ...
}
```

zbjornson commented on src/bmp/BMPParser.cc:
```diff
@@ -0,0 +1,403 @@
+  EX(fhSig != "BM", temp); // BM
+
+  // Length of the file should be equal to `len`
+  E(U4() != (uint32_t)len, "inconsistent file size");
```
Minor: prefer `static_cast` over C-style.

zbjornson commented on src/bmp/BMPParser.cc:
```diff
@@ -0,0 +1,403 @@
+  }
+
+  // Ensure that all image data is present
+  E(ptr - data + imgdSize > len, "no enough image data");
```
minor: typo (no -> not)

LinusU commented:
Haven't looked to closely at the code, but feel free to merge if you think it's good @zbjornson 👍

chearon commented:
Neither have I but it looks good and I must say it's quite a relief to not have to update prebuilds for this!
